chainstate/tx-verifier/src/transaction_verifier/input_check/mod.rs

```diff
@@ -0,0 +1,458 @@
+            } => current_best.block_id(),
+        };
+
+        Self::custom(
+            chain_config,
+            storage,
+            tip,
+            spending_time,
+            tx_source.expected_block_height(),
+            core_ctx,
+        )
+    }
+}
+
+struct InputVerifyContextTimelock<'a, S> {
+    ctx: &'a VerifyContextTimelock<'a, S>,
+    input_num: usize,
+}
+
+impl<'a, S> InputVerifyContextTimelock<'a, S> {
+    fn new(ctx: &'a VerifyContextTimelock<'a, S>, input_num: usize) -> Self {
+        Self { ctx, input_num }
+    }
+
+    fn info(&self) -> &InputInfo {
+        self.ctx.core_ctx.input_data(self.input_num).input_info()
+    }
+}
+
+impl<S: TransactionVerifierStorageRef> TimelockContext for InputVerifyContextTimelock<'_, S> {
+    type Error = ConnectTransactionError;
+
+    fn spending_height(&self) -> BlockHeight {
+        self.ctx.spending_height
+    }
+
+    fn spending_time(&self) -> BlockTimestamp {
+        self.ctx.spending_time
+    }
+
+    fn source_height(&self) -> Result<BlockHeight, Self::Error> {
+        match self.info() {
+            InputInfo::Utxo { outpoint: _, utxo } => match utxo.source() {
+                utxo::UtxoSource::Blockchain(height) => Ok(*height),
+                utxo::UtxoSource::Mempool => Ok(self.ctx.spending_height),
+            },
+            InputInfo::Account { .. } | InputInfo::AccountCommand { .. } => {
+                Err(ConnectTransactionError::TimelockedAccount)
+            }
+        }
+    }
+
+    fn source_time(&self) -> Result<BlockTimestamp, Self::Error> {
+        match self.info() {
+            InputInfo::Utxo { outpoint: _, utxo } => match utxo.source() {
+                utxo::UtxoSource::Blockchain(height) => {
+                    let block_index_getter = |db_tx: &S, _: &ChainConfig, id: &Id<GenBlock>| {
+                        db_tx.get_gen_block_index(id)
+                    };
+
+                    let source_block_index = block_index_ancestor_getter(
+                        block_index_getter,
+                        self.ctx.storage,
+                        self.ctx.chain_config,
+                        (&self.ctx.tip).into(),
+                        *height,
+                    )
+                    .map_err(|e| {
+                        ConnectTransactionError::InvariantErrorHeaderCouldNotBeLoadedFromHeight(
+                            e, *height,
+                        )
+                    })?;
+
+                    Ok(source_block_index.block_timestamp())
+                }
+                utxo::UtxoSource::Mempool => Ok(self.ctx.spending_time),
+            },
+            InputInfo::Account { .. } | InputInfo::AccountCommand { .. } => {
+                Err(ConnectTransactionError::TimelockedAccount)
+            }
+        }
+    }
+}
+
+struct VerifyContextFull<'a, T, S> {
+    // Here we need more information about the transaction
+    transaction: &'a T,
+    spent_utxos: Vec<Option<&'a TxOutput>>,
```

Comment:
I would question whether we need these in here. After all, when we do the verification, we are including the utxo source (InputInfo) anyway. Maybe I'm missing something, I'm now sure.

Reply:
> I would question whether we need these in here. After all, when we do the verification, we are including the utxo source (InputInfo) anyway. Maybe I'm missing something, I'm not sure.

The signature checking function expects the list of inputs to be committed to and it takes it in this format.